src/moves.rs

```diff
@@ -0,0 +1,204 @@
+/*
+ * Released under the terms of the Apache 2.0 license with LLVM
+ * exception. See `LICENSE` for details.
+ */
+
+use crate::Allocation;
+use smallvec::{smallvec, SmallVec};
+
+pub type MoveVec<T> = SmallVec<[(Allocation, Allocation, T); 16]>;
+
+/// A `ParallelMoves` represents a list of alloc-to-alloc moves that
+/// must happen in parallel -- i.e., all reads of sources semantically
+/// happen before all writes of destinations, and destinations are
+/// allowed to overwrite sources. It can compute a list of sequential
+/// moves that will produce the equivalent data movement, possibly
+/// using a scratch register if one is necessary.
+pub struct ParallelMoves<T: Clone + Copy + Default> {
+    parallel_moves: MoveVec<T>,
+    scratch: Allocation,
+}
+
+impl<T: Clone + Copy + Default> ParallelMoves<T> {
+    pub fn new(scratch: Allocation) -> Self {
+        Self {
+            parallel_moves: smallvec![],
+            scratch,
+        }
+    }
+
+    pub fn add(&mut self, from: Allocation, to: Allocation, t: T) {
+        self.parallel_moves.push((from, to, t));
+    }
+
+    fn sources_overlap_dests(&self) -> bool {
+        // Assumes `parallel_moves` has already been sorted in `resolve()` below.
+        for &(_, dst, _) in &self.parallel_moves {
+            if self
+                .parallel_moves
+                .binary_search_by_key(&dst, |&(src, _, _)| src)
+                .is_ok()
+            {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn resolve(mut self) -> MoveVec<T> {
+        // Easy case: zero or one move. Just return our vec.
+        if self.parallel_moves.len() <= 1 {
+            return self.parallel_moves;
+        }
+
+        // Sort moves by source so that we can efficiently test for
+        // presence.
+        self.parallel_moves.sort_by_key(|&(src, dst, _)| (src, dst));
+
+        // Do any dests overlap sources? If not, we can also just
+        // return the list.
+        if !self.sources_overlap_dests() {
+            return self.parallel_moves;
+        }
+
+        // General case: some moves overwrite dests that other moves
+        // read as sources. We'll use a general algorithm.
+        //
+        // *Important property*: because we expect that each register
+        // has only one writer (otherwise the effect of the parallel
+        // move is undefined), each move can only block one other move
+        // (with its one source corresponding to the one writer of
+        // that source). Thus, we *can only have simple cycles*: there
```

Comment:
What does 'simple' mean here?  I'm imagining that a simple cycle is an SCC in which there is only one possible path from each node back to itself.  Eg A->B, B->C, C->D, D->A is a simple cycle, but adding A->C to that would make it be
not a simple cycle, because then there are two paths from A back to A: ABCDA and ACDA.

Reply:
Indeed, clarified the text here; thanks!